### ISSUE TITLE
PartialLinearFluid - reference_d sometimes used instead of density

### ISSUE THREAD

**Issue by modelica-trac-importer** [modelica/Modelica#110]
**Reported by otter on 8 Oct 2008 09:22 UTC**
In Modelica.Media.Interfaces.PartialLinearFluid property calculations are sometimes performed with "reference_d", although "density(p,T)" should have been used. This is e.g. the case in:

 * setState_phX
 * setState_dTX
 * setState_psX
 * T_ph
 * T_ps

----

Migrated-From: https://trac.modelica.org/Modelica/ticket/110

**Comment by modelica-trac-importer** [comment]
**Comment by hubertus on 8 Oct 2008 21:38 UTC**
That was done on purpose: a feature, not a bug, one could say. Density varies little or very little, and using a fixed density keeps many "Thermodynamic Jacobians" of state changes konstant, so that equations can be solved at compile time. There are many variants overall which properties should be modeled constant, linear, or higher order and non-linear in some way. This particular choice was on purpose, but there is no easy way to name the properties in Detail, like LinarCpConstantDensityFluid or whatever.

**Comment by modelica-trac-importer** [comment]
**Comment by dietmarw on 9 Oct 2008 06:28 UTC**
Others might wonder about this too. So making this ticket public makes it possible to find Hubertus' explanation.

**Comment by modelica-trac-importer** [comment]
**Comment by otter on 9 Oct 2008 07:16 UTC**
It is fine to make such assumptions. However, it would be better if the assumptions could be made consistent. So, either all functions or none of the functions should make the assumption that density is constant. If in some applications the varying density value is needed, then this could be either handled with an additional argument to the functions or a special new function (densityDetailed(..)).

Note, at least the documentation is currently wrong, because it is stated that setting "constantJacobian = false" would use the varying density in all functions. This is not true.

The issue becomes important if inverses of functions are utilized by a tool and the inverse gives a different result as the direction function evaluation. This is actually the case for the PartialLinearFluid medium:

```
state = setState_phX(p,h1,X)
h2    = specificEnthalpy(state)
T     = T_ph(p,h2)
```

since h1 <> h2 and state.T <> T, although equality should hold. The reason is that

 * setState_phX computes state.T always with constant density
 * specificEnhalpy computes the specific enthalpy always with non-constant density
 * T_ph computes T always with constant density.

This is an actual problem, because Sven Erik, Hans, Hilding and me are working on medium propagation over ports in order to remove a scalar non-linear equation system for (p,T) media at every 1:1 connection. This problem can be nicely solved if the tool gets the information that function h_pT() is the inverse of T_ph(), and a good implementation will have a "check" to verify that the statement of the user is correct (that one function is the inverse of the other one). Such a check would fail with the implementation above.

One could proceed in the following way:

 1. Introduce a new function "h_pT(..)" that uses the implementation of setState_phX (i.e., constant density).

 2. Replace the implementation of setState_phX by a call to h_pT(..)

 3. Define that h_pT is the inverse of T_ph (this is consistent because all functions use constant density).

However, what is not nice is that enthalpy(state) will give a different result as h_pT(..).

I wonder whether the "constant Jacobian" feature is really measurable. Usually, if nonlinear equation systems appear there are other parts that do not lead to a constant Jacobian and then this does not matter. A clean implementation of this medium model would be to have a constant that signals whether constant or varying density is used in all functions.

**Comment by modelica-trac-importer** [comment]
**Comment by hubertus on 21 Oct 2008 08:07 UTC**
Fixed in commit #1175. Used reference_d instead of density(state) in order to make inverses exact w.r.t. forward functions.

**Comment by modelica-trac-importer** [comment]
**Comment by dietmarw on 21 Oct 2008 10:43 UTC**
Fixed in commit 7a4144df555aa4f1cfd5f69b4e3579a26d14b599. Used reference_d instead of density(state) in order to make inverses exact w.r.t. forward functions.

(use #XXX for tickets and rXXX for commits :) )
